qubesmanager/backup.py

```diff
@@ -310,65 +323,56 @@ def current_page_changed(self, page_id): # pylint: disable=unused-argument
-            thread = threading.Thread(
-                target=self.__do_backup__,
-                args=(self.thread_monitor,))
-            thread.daemon = True
-            thread.start()
-
-            while not self.thread_monitor.is_finished():
-                self.qt_app.processEvents()
-                time.sleep(0.1)
-
-            if not self.thread_monitor.success:
-                if self.canceled:
-                    self.progress_status.setText(
-                        self.tr(
-                            "Backup aborted. "
-                            "Temporary file may be left at backup location."))
-                else:
-                    self.progress_status.setText(self.tr("Backup error."))
-                    QtGui.QMessageBox.warning(
-                        self, self.tr("Backup error!"),
-                        self.tr("ERROR: {}").format(
-                            self.thread_monitor.error_msg))
-            else:
-                self.progress_bar.setMaximum(100)
-                self.progress_bar.setValue(100)
-                self.progress_status.setText(self.tr("Backup finished."))
+
+            vm = self.qubes_app.domains[
+                self.appvm_combobox.currentText()]
+
+            self.thread = BackupThread(vm)
+            self.thread.finished.connect(self.backup_finished)
+            self.thread.start()
+
+        signal.signal(signal.SIGCHLD, old_sigchld_handler)
+
+    def backup_finished(self):
+        if self.thread.msg:
+            self.progress_status.setText(self.tr("Backup error."))
+            QtGui.QMessageBox.warning(
+                self, self.tr("Backup error!"),
+                self.tr("ERROR: {}").format(
+                    self.thread.msg))
+        else:
+            self.progress_bar.setValue(100)
+            self.progress_status.setText(self.tr("Backup finished."))
+
             if self.showFileDialog.isChecked():
                 orig_text = self.progress_status.text
                 self.progress_status.setText(
                     orig_text + self.tr(
                         " Please unmount your backup volume and cancel "
                         "the file selection dialog."))
                 backup_utils.select_path_button_clicked(self, False, True)
+
             self.button(self.CancelButton).setEnabled(False)
             self.button(self.FinishButton).setEnabled(True)
             self.showFileDialog.setEnabled(False)
             self.cleanup_temporary_files()
 
             # turn off only when backup was successful
-            if self.thread_monitor.success and \
-                    self.turn_off_checkbox.isChecked():
+            if self.turn_off_checkbox.isChecked():
                 os.system('systemctl poweroff')
 
-        signal.signal(signal.SIGCHLD, old_sigchld_handler)
-
     def reject(self):
         if self.currentPage() is self.commit_page:
-            self.canceled = True
+            self.thread.terminate()
```

Comment:
Better call `admin.backup.Cancel` (the line below) and wait until it really is canceled - the `admin.backup.Execute` method should exit with exception, so there is no need to cancel that thread.